File: skills/continuum-tracker/scripts/ct.mjs
#!/usr/bin/env node
/**
 * Continuum Tracker read-only API helper for the continuum-tracker skill.
 *
 * Node built-ins only (no npm install; requires Node 18+ for global fetch).
 * 1:1 port of ct.py — same commands, flags, and output. Use whichever
 * runtime is available on the system.
 *
 * Auth: CONTINUUM_API_KEY env var, else CONTINUUM_API_KEY=... in ./.env,
 * else the command exits asking for it. The key is never printed or stored.
 *
 * Usage:
 *   node ct.mjs me
 *   node ct.mjs projects
 *   node ct.mjs project   --project PID
 *   node ct.mjs signals   --project PID [--top N] [--search T] [--category C] [--saved] [--all]
 *   node ct.mjs feedbacks --project PID [--search T] [--limit N] [--all]
 *   node ct.mjs feedback  --project PID --id FID
 *   node ct.mjs signal    --project PID --id SID
 *   node ct.mjs signal-feedbacks --project PID --id SID [--all]
 *   node ct.mjs painpoints --project PID (--feedback FID | --signal SID) [--all]
 *   node ct.mjs signal-market --project PID --id SID
 *
 * Add --json to any command for the raw API response.
 */
import fs from "node:fs";
import process from "node:process";

const OFFICIAL_HOST = "app.continuumtracker.com";

/**
 * Resolve the API base, refusing anywhere the key must not be sent.
 * CONTINUUM_API_BASE is an env-var override, so it is an untrusted input: every
 * request carries `Authorization: Bearer <key>`. Guard it before the key moves.
 */
function resolveBase() {
  const raw = (process.env.CONTINUUM_API_BASE || `https://${OFFICIAL_HOST}/api`).trim().replace(/\/+$/, "");
  let parsed;
  try {
    parsed = new URL(raw);
  } catch {
    console.error(`CONTINUUM_API_BASE must be an http(s) URL with a host; got: ${raw}`);
    process.exit(1);
  }
  const host = parsed.hostname.toLowerCase();
  const isLocal = ["localhost", "127.0.0.1", "::1"].includes(host);

  if (!["http:", "https:"].includes(parsed.protocol) || !host) {
    console.error(`CONTINUUM_API_BASE must be an http(s) URL with a host; got: ${raw}`);
    process.exit(1);
  }
  if (parsed.protocol !== "https:" && !isLocal) {
    console.error(
      `Refusing to send your API key over plain HTTP to a remote host (${host}). ` +
        "Use https:// in CONTINUUM_API_BASE.",
    );
    process.exit(1);
  }
  if (host !== OFFICIAL_HOST && !isLocal) {
    console.error(
      `warning: CONTINUUM_API_BASE points at ${host} - your Continuum Tracker ` +
        "API key will be sent there, not to the official API.",
    );
  }
  return raw;
}

const BASE = resolveBase();
const TIMEOUT = 30000;
const MAX_RETRIES = 3;
let KEY = "";

// --------------------------------------------------------------------------- key
function loadKey() {
  const env = (process.env.CONTINUUM_API_KEY || "").trim();
  if (env) return env;
  try {
    for (const line of fs.readFileSync(".env", "utf-8").split(/\r?\n/)) {
      const t = line.trim();
      if (t.startsWith("CONTINUUM_API_KEY=")) {
        return t.slice("CONTINUUM_API_KEY=".length).trim().replace(/^['"]|['"]$/g, "");
      }
    }
  } catch { /* no .env */ }
  die(
    "No API key found. Set CONTINUUM_API_KEY in the environment or in a .env " +
    "file in this directory. Generate a key in the web app at /settings/access."
  );
}

function die(msg, code = 1) {
  process.stderr.write(`error: ${msg}\n`);
  process.exit(code);
}

// -------------------------------------------------------------------------- http
function buildQuery(params) {
  const qs = new URLSearchParams();
  for (const [k, v] of Object.entries(params || {})) {
    if (v === null || v === undefined) continue;
    if (Array.isArray(v)) v.forEach((x) => x != null && qs.append(k, String(x)));
    else qs.append(k, String(v));
  }
  const s = qs.toString();
  return s ? `?${s}` : "";
}

async function request(path, params) {
  const url = `${BASE}${path}${buildQuery(params)}`;
  const headers = { Authorization: `Bearer ${KEY}`, Accept: "application/json" };
  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    let resp;
    try {
      resp = await fetch(url, { method: "GET", headers, signal: AbortSignal.timeout(TIMEOUT) });
    } catch (e) {
      die(`network error: ${e.message || e}`, 3);
    }
    const body = await resp.text();
    if (resp.ok) return body ? JSON.parse(body) : {};
    if (resp.status === 429 && attempt < MAX_RETRIES - 1) {
      const wait = retryAfter(resp, body);
      process.stderr.write(`rate limited, retrying in ${wait}s...\n`);
      await sleep(wait * 1000);
      continue;
    }
    die(httpErrorMsg(resp.status, body), 2);
  }
  die("exhausted retries", 2);
}

function retryAfter(resp, body) {
  const hdr = resp.headers.get("retry-after");
  if (hdr && /^\d+$/.test(hdr)) return parseInt(hdr, 10);
  try { return parseInt(JSON.parse(body)?.error?.retry_after, 10) || 5; } catch { return 5; }
}

function httpErrorMsg(status, body) {
  try {
    const err = JSON.parse(body)?.error || {};
    const code = err.code || "";
    const message = err.message || body;
    const prefix = { 401: "unauthorized (check API key)", 403: "forbidden",
      404: "not found", 422: "validation error" }[status] || `HTTP ${status}`;
    return `${prefix}: ${code ? `[${code}] ` : ""}${message}`.trim();
  } catch {
    return `HTTP ${status}: ${body.slice(0, 300)}`;
  }
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function paginate(path, params, followAll) {
  params = { page: 1, ...params };
  const items = [];
  for (;;) {
    const data = await request(path, params);
    items.push(...(data.items || []));
    const pg = data.pagination || {};
    if (!followAll || params.page >= (pg.total_pages || 1)) return items;
    params.page += 1;
  }
}

// ------------------------------------------------------------------------ render
function trunc(text, n = 90) {
  if (!text) return "";
  text = String(text).split(/\s+/).join(" ");
  return text.length <= n ? text : text.slice(0, n - 1) + "…";
}
const outJson = (obj) => console.log(JSON.stringify(obj, null, 2));
const pad3 = (n) => String(n ?? 0).padStart(3);

// ----------------------------------------------------------------------- commands
async function cmdMe(a) {
  const d = await request("/v1/me");
  if (a.json) return outJson(d);
  console.log(`${d.email || "?"}  (id ${d.id || "?"})`);
  console.log(`default org: ${d.default_organization_id || "-"}`);
}

async function cmdProjects(a) {
  const items = await paginate("/v1/projects", { limit: 100 }, true);
  if (a.json) return outJson(items);
  if (!items.length) return console.log("no projects");
  for (const p of items) {
    console.log(`- ${p.name || "(unnamed)"}  [${p.id}]${p.is_default ? " *default" : ""}`);
    for (const f of ["mission", "vision", "north_star"]) {
      if (p[f]) console.log(`    ${f}: ${trunc(p[f], 120)}`);
    }
  }
}

async function cmdProject(a) {
  const d = await request(`/v1/projects/${a.project}`);
  if (a.json) return outJson(d);
  console.log(`${d.name || "(unnamed)"}  [${d.id}]${d.is_default ? "  *default" : ""}`);
  if (d.url) console.log(`url: ${d.url}`);
  for (const f of ["mission", "vision", "north_star", "similar_companies"]) {
    if (d[f]) console.log(`\n${f.replace(/_/g, " ").toUpperCase()}:\n${d[f]}`);
  }
}

async function cmdSignals(a) {
  const params = {
    limit: a.limit || 100,
    search: a.search,
    category: a.category || null,
    is_saved: a.saved ? "true" : null,
  };
  let items = await paginate(`/v1/projects/${a.project}/signals`, params, a.all);
  items.sort((x, y) => (y.feedback_count || 0) - (x.feedback_count || 0));
  if (a.top) items = items.slice(0, a.top);
  if (a.json) return outJson(items);
  if (!items.length) return console.log("no signals");
  for (const s of items) {
    console.log(`[${pad3(s.feedback_count)} fb] ${s.name || "(unnamed)"}  (${s.category})  [${s.id}]`);
    if (s.pain_point) console.log(`        pain: ${trunc(s.pain_point)}`);
    if (s.market_opportunity) console.log(`        market: ${trunc(s.market_opportunity)}`);
  }
}

async function cmdFeedbacks(a) {
  const params = { limit: a.limit || 100, search: a.search, sort_by: "created_at", sort_dir: "desc" };
  const items = await paginate(`/v1/projects/${a.project}/feedbacks`, params, a.all);
  if (a.json) return outJson(items);
  if (!items.length) return console.log("no feedbacks");
  for (const f of items) {
    console.log(`- ${f.name || "(untitled)"}  by ${f.author || "?"}  [${f.id}]`);
    console.log(`    ${trunc(f.feedback_original)}`);
  }
}

async function cmdFeedback(a) {
  const d = await request(`/v1/projects/${a.project}/feedbacks/${a.id}`, { include: "painpoint" });
  if (a.json) return outJson(d);
  console.log(`${d.name || "(untitled)"}  [${d.id}]`);
  console.log(`author: ${d.author || "?"} (${d.author_type || "-"})  source: ${d.source}  status: ${d.status}`);
  console.log(`\n${d.feedback_original || ""}\n`);
  if (d.feedback_processed) console.log(`processed: ${d.feedback_processed}`);
  for (const pp of d.painpoints || []) {
    console.log(`  - painpoint: ${trunc(pp.painpoint_original || pp.pain_point || pp.name)}`);
    if (pp.painpoint_processed) console.log(`      -> ${trunc(pp.painpoint_processed, 200)}`);
  }
}

async function cmdSignal(a) {
  const d = await request(`/v1/projects/${a.project}/signals/${a.id}`);
  if (a.json) return outJson(d);
  console.log(`${d.name}  (${d.category})  [${d.feedback_count || 0} feedbacks]`);
  for (const f of ["pain_point", "user_story", "market_opportunity"]) {
    if (d[f]) console.log(`  ${f}: ${trunc(d[f], 140)}`);
  }
}

async function cmdSignalFeedbacks(a) {
  const params = { limit: a.limit || 100, include: "painpoint" };
  const items = await paginate(`/v1/projects/${a.project}/signals/${a.id}/feedbacks`, params, a.all);
  if (a.json) return outJson(items);
  if (!items.length) return console.log("no feedbacks for this signal");
  for (const f of items) {
    console.log(`- ${f.name || "(untitled)"}  [${f.id}]`);
    console.log(`    ${trunc(f.feedback_original)}`);
  }
}

async function cmdPainpoints(a) {
  let path;
  if (a.feedback) path = `/v1/projects/${a.project}/feedbacks/${a.feedback}/painpoints`;
  else if (a.signal) path = `/v1/projects/${a.project}/signals/${a.signal}/painpoints`;
  else die("painpoints needs --feedback FID or --signal SID");
  const items = await paginate(path, { limit: a.limit || 50 }, a.all);
  if (a.json) return outJson(items);
  if (!items.length) return console.log("no painpoints");
  for (const pp of items) {
    const quote = pp.painpoint_original || pp.pain_point || pp.name;
    console.log(`- "${trunc(quote, 300)}"  [${pp.id}]`);
    if (pp.painpoint_processed) console.log(`    -> ${trunc(pp.painpoint_processed, 200)}`);
    if (pp.feedback_id) console.log(`    feedback: ${pp.feedback_id}`);
  }
}

async function cmdSignalMarket(a) {
  const d = await request(`/v1/projects/${a.project}/signals/${a.id}/market-research`);
  if (a.json) return outJson(d);
  const items = d.items || [];
  if (!items.length) return console.log("no market research for this signal");
  console.log(`${items.length} market sources\n`);
  for (const r of items) {
    const src = [r.company_name, r.product_name].filter(Boolean).join(" / ");
    console.log(r.score != null
      ? `[${r.score.toFixed(3)}] ${src || "(unknown source)"}`
      : `  ${src || "(unknown source)"}`);
    console.log(`    ${trunc(r.text, 120)}`);
    if (r.business_description) console.log(`    about: ${trunc(r.business_description, 100)}`);
    if (r.released_at) console.log(`    released: ${r.released_at}`);
  }
}

// --------------------------------------------------------------------------- cli
const COMMANDS = {
  me: { fn: cmdMe },
  projects: { fn: cmdProjects },
  project: { fn: cmdProject, project: true },
  signals: { fn: cmdSignals, project: true,
    flags: { top: "int", limit: "int", search: "str", category: "multi", saved: "bool", all: "bool" } },
  feedbacks: { fn: cmdFeedbacks, project: true,
    flags: { search: "str", limit: "int", all: "bool" } },
  feedback: { fn: cmdFeedback, project: true, flags: { id: "str!" } },
  signal: { fn: cmdSignal, project: true, flags: { id: "str!" } },
  "signal-feedbacks": { fn: cmdSignalFeedbacks, project: true,
    flags: { id: "str!", limit: "int", all: "bool" } },
  painpoints: { fn: cmdPainpoints, project: true,
    flags: { feedback: "str", signal: "str", limit: "int", all: "bool" } },
  "signal-market": { fn: cmdSignalMarket, project: true, flags: { id: "str!" } },
};

function parse(command, argv) {
  const spec = COMMANDS[command];
  const flags = { ...(spec.flags || {}) };
  if (spec.project) flags.project = "str!";
  flags.json = "bool";
  const out = {};
  for (let i = 0; i < argv.length; i++) {
    const t = argv[i];
    if (!t.startsWith("--")) die(`unexpected argument: ${t}`, 2);
    const name = t.slice(2);
    const kind = flags[name];
    if (!kind) die(`${command}: unknown option --${name}`, 2);
    if (kind === "bool") { out[name] = true; continue; }
    const val = argv[++i];
    if (val === undefined) die(`${command}: --${name} needs a value`, 2);
    if (kind === "multi") (out[name] ||= []).push(val);
    else if (kind.startsWith("int")) out[name] = parseInt(val, 10);
    else out[name] = val;
  }
  for (const [name, kind] of Object.entries(flags)) {
    if (typeof kind === "string" && kind.endsWith("!") && out[name] === undefined) {
      die(`${command}: the following argument is required: --${name}`, 2);
    }
  }
  return out;
}

async function main() {
  const [command, ...rest] = process.argv.slice(2);
  if (!command || command === "-h" || command === "--help") {
    console.log(`usage: ct.mjs {${Object.keys(COMMANDS).join(",")}} [options]   (add --json for raw output)`);
    process.exit(command ? 0 : 2);
  }
  if (!COMMANDS[command]) die(`unknown command: ${command}`, 2);
  const args = parse(command, rest);
  KEY = loadKey();
  await COMMANDS[command].fn(args);
}

main().catch((e) => die(e.message || String(e), 1));

File: skills/continuum-tracker/scripts/ct.py
#!/usr/bin/env python3
"""Continuum Tracker read-only API helper for the continuum-tracker skill.

Standard-library only (no pip install). Resolves the API key, handles
pagination and 429 back-off, and prints compact human-readable output
(or raw JSON with --json).

Auth: CONTINUUM_API_KEY env var, else CONTINUUM_API_KEY=... in ./.env,
else the command exits asking for it. The key is never printed or stored.

Usage:
  ct.py me
  ct.py projects
  ct.py project   --project PID
  ct.py signals   --project PID [--top N] [--search T] [--category C] [--saved] [--all]
  ct.py feedbacks --project PID [--search T] [--limit N] [--all]
  ct.py feedback  --project PID --id FID
  ct.py signal    --project PID --id SID
  ct.py signal-feedbacks --project PID --id SID [--all]
  ct.py painpoints --project PID (--feedback FID | --signal SID) [--all]
  ct.py signal-market --project PID --id SID

Add --json to any command for the raw API response.
"""
from __future__ import annotations

import argparse
import json
import os
import sys
import time
import urllib.error
import urllib.parse
import urllib.request

try:  # ensure non-ASCII (em-dashes, accents, ellipsis) prints cleanly on Windows
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
except Exception:
    pass

OFFICIAL_HOST = "app.continuumtracker.com"


def _resolve_base() -> str:
    """Resolve the API base, refusing anywhere the key must not be sent.

    CONTINUUM_API_BASE is an env-var override, so it is an untrusted input: every
    request carries `Authorization: Bearer <key>`. Guard it before the key moves.
    """
    raw = os.environ.get("CONTINUUM_API_BASE", f"https://{OFFICIAL_HOST}/api").strip().rstrip("/")
    parsed = urllib.parse.urlparse(raw)
    host = (parsed.hostname or "").lower()
    is_local = host in ("localhost", "127.0.0.1", "::1")

    if parsed.scheme not in ("http", "https") or not host:
        sys.exit(f"CONTINUUM_API_BASE must be an http(s) URL with a host; got: {raw!r}")
    if parsed.scheme != "https" and not is_local:
        sys.exit(
            "Refusing to send your API key over plain HTTP to a remote host "
            f"({host}). Use https:// in CONTINUUM_API_BASE."
        )
    if host != OFFICIAL_HOST and not is_local:
        print(
            f"warning: CONTINUUM_API_BASE points at {host} - your Continuum Tracker "
            "API key will be sent there, not to the official API.",
            file=sys.stderr,
        )
    return raw


BASE = _resolve_base()
TIMEOUT = 30
MAX_RETRIES = 3


# --------------------------------------------------------------------------- key
def load_key() -> str:
    key = os.environ.get("CONTINUUM_API_KEY", "").strip()
    if key:
        return key
    try:
        with open(".env", encoding="utf-8") as fh:
            for line in fh:
                line = line.strip()
                if line.startswith("CONTINUUM_API_KEY="):
                    return line.split("=", 1)[1].strip().strip("\"'")
    except OSError:
        pass
    die(
        "No API key found. Set CONTINUUM_API_KEY in the environment or in a .env "
        "file in this directory. Generate a key in the web app at /settings/access."
    )


def die(msg: str, code: int = 1) -> "None":
    print(f"error: {msg}", file=sys.stderr)
    sys.exit(code)


# -------------------------------------------------------------------------- http
def request(path: str, params: dict | None = None) -> dict:
    url = f"{BASE}{path}"
    if params:
        clean = {k: v for k, v in params.items() if v is not None}
        if clean:
            # doseq=True lets list values repeat the param (e.g. category)
            url += "?" + urllib.parse.urlencode(clean, doseq=True)
    headers = {"Authorization": f"Bearer {KEY}", "Accept": "application/json"}
    for attempt in range(MAX_RETRIES):
        req = urllib.request.Request(url, headers=headers, method="GET")
        try:
            with urllib.request.urlopen(req, timeout=TIMEOUT) as resp:
                return json.loads(resp.read().decode("utf-8") or "{}")
        except urllib.error.HTTPError as exc:
            body = exc.read().decode("utf-8", "replace")
            if exc.code == 429 and attempt < MAX_RETRIES - 1:
                wait = _retry_after(exc, body)
                print(f"rate limited, retrying in {wait}s...", file=sys.stderr)
                time.sleep(wait)
                continue
            die(_http_error_msg(exc.code, body), code=2)
        except urllib.error.URLError as exc:
            die(f"network error: {exc.reason}", code=3)
    die("exhausted retries", code=2)


def _retry_after(exc: urllib.error.HTTPError, body: str) -> int:
    hdr = exc.headers.get("Retry-After")
    if hdr and hdr.isdigit():
        return int(hdr)
    try:
        return int(json.loads(body).get("error", {}).get("retry_after") or 5)
    except Exception:
        return 5


def _http_error_msg(status: int, body: str) -> str:
    try:
        err = json.loads(body).get("error", {})
        code = err.get("code", "")
        message = err.get("message", body)
        prefix = {401: "unauthorized (check API key)", 403: "forbidden",
                  404: "not found", 422: "validation error"}.get(status, f"HTTP {status}")
        return f"{prefix}: {('[' + code + '] ') if code else ''}{message}".strip()
    except Exception:
        return f"HTTP {status}: {body[:300]}"


def paginate(path: str, params: dict, follow_all: bool) -> list:
    """Return list of items; follow pages when follow_all is set."""
    params = dict(params)
    params.setdefault("page", 1)
    items: list = []
    while True:
        data = request(path, params)
        items.extend(data.get("items", []))
        pg = data.get("pagination", {})
        if not follow_all or params["page"] >= pg.get("total_pages", 1):
            return items
        params["page"] += 1


# ------------------------------------------------------------------------ render
def trunc(text, n: int = 90) -> str:
    if not text:
        return ""
    text = " ".join(str(text).split())
    return text if len(text) <= n else text[: n - 1] + "…"


def out_json(obj) -> None:
    print(json.dumps(obj, indent=2, ensure_ascii=False))


# ----------------------------------------------------------------------- commands
def cmd_me(args) -> None:
    data = request("/v1/me")
    if args.json:
        return out_json(data)
    print(f"{data.get('email', '?')}  (id {data.get('id', '?')})")
    print(f"default org: {data.get('default_organization_id', '-')}")


def cmd_projects(args) -> None:
    items = paginate("/v1/projects", {"limit": 100}, follow_all=True)
    if args.json:
        return out_json(items)
    if not items:
        return print("no projects")
    for p in items:
        star = " *default" if p.get("is_default") else ""
        print(f"- {p.get('name') or '(unnamed)'}  [{p.get('id')}]{star}")
        for field in ("mission", "vision", "north_star"):
            if p.get(field):
                print(f"    {field}: {trunc(p[field], 120)}")


def cmd_project(args) -> None:
    d = request(f"/v1/projects/{args.project}")
    if args.json:
        return out_json(d)
    star = "  *default" if d.get("is_default") else ""
    print(f"{d.get('name') or '(unnamed)'}  [{d.get('id')}]{star}")
    if d.get("url"):
        print(f"url: {d['url']}")
    for field in ("mission", "vision", "north_star", "similar_companies"):
        if d.get(field):
            print(f"\n{field.replace('_', ' ').upper()}:\n{d[field]}")


def cmd_signals(args) -> None:
    params = {
        "limit": args.limit or 100,
        "search": args.search,
        "category": args.category or None,
        "is_saved": "true" if args.saved else None,
    }
    items = paginate(f"/v1/projects/{args.project}/signals", params, follow_all=args.all)
    # rank by evidence (feedback_count desc) — the API only sorts by created_at/name
    items.sort(key=lambda s: s.get("feedback_count", 0), reverse=True)
    if args.top:
        items = items[: args.top]
    if args.json:
        return out_json(items)
    if not items:
        return print("no signals")
    for s in items:
        print(f"[{s.get('feedback_count', 0):>3} fb] {s.get('name') or '(unnamed)'}  "
              f"({s.get('category')})  [{s.get('id')}]")
        if s.get("pain_point"):
            print(f"        pain: {trunc(s['pain_point'])}")
        if s.get("market_opportunity"):
            print(f"        market: {trunc(s['market_opportunity'])}")


def cmd_feedbacks(args) -> None:
    params = {
        "limit": args.limit or 100,
        "search": args.search,
        "sort_by": "created_at",
        "sort_dir": "desc",
    }
    items = paginate(f"/v1/projects/{args.project}/feedbacks", params, follow_all=args.all)
    if args.json:
        return out_json(items)
    if not items:
        return print("no feedbacks")
    for f in items:
        who = f.get("author") or "?"
        print(f"- {f.get('name') or '(untitled)'}  by {who}  [{f.get('id')}]")
        print(f"    {trunc(f.get('feedback_original'))}")


def cmd_feedback(args) -> None:
    data = request(f"/v1/projects/{args.project}/feedbacks/{args.id}",
                   {"include": "painpoint"})
    if args.json:
        return out_json(data)
    print(f"{data.get('name') or '(untitled)'}  [{data.get('id')}]")
    print(f"author: {data.get('author') or '?'} ({data.get('author_type') or '-'})  "
          f"source: {data.get('source')}  status: {data.get('status')}")
    print(f"\n{data.get('feedback_original', '')}\n")
    if data.get("feedback_processed"):
        print(f"processed: {data['feedback_processed']}")
    for pp in data.get("painpoints", []) or []:
        quote = pp.get("painpoint_original") or pp.get("pain_point") or pp.get("name")
        print(f"  - painpoint: {trunc(quote)}")
        if pp.get("painpoint_processed"):
            print(f"      -> {trunc(pp['painpoint_processed'], 200)}")


def cmd_signal(args) -> None:
    data = request(f"/v1/projects/{args.project}/signals/{args.id}")
    if args.json:
        return out_json(data)
    print(f"{data.get('name')}  ({data.get('category')})  "
          f"[{data.get('feedback_count', 0)} feedbacks]")
    for field in ("pain_point", "user_story", "market_opportunity"):
        if data.get(field):
            print(f"  {field}: {trunc(data[field], 140)}")


def cmd_signal_feedbacks(args) -> None:
    params = {"limit": args.limit or 100, "include": "painpoint"}
    items = paginate(f"/v1/projects/{args.project}/signals/{args.id}/feedbacks",
                     params, follow_all=args.all)
    if args.json:
        return out_json(items)
    if not items:
        return print("no feedbacks for this signal")
    for f in items:
        print(f"- {f.get('name') or '(untitled)'}  [{f.get('id')}]")
        print(f"    {trunc(f.get('feedback_original'))}")


def cmd_painpoints(args) -> None:
    if args.feedback:
        path = f"/v1/projects/{args.project}/feedbacks/{args.feedback}/painpoints"
    elif args.signal:
        path = f"/v1/projects/{args.project}/signals/{args.signal}/painpoints"
    else:
        die("painpoints needs --feedback FID or --signal SID")
    items = paginate(path, {"limit": args.limit or 50}, follow_all=args.all)
    if args.json:
        return out_json(items)
    if not items:
        return print("no painpoints")
    for pp in items:
        quote = pp.get("painpoint_original") or pp.get("pain_point") or pp.get("name")
        print(f"- \"{trunc(quote, 300)}\"  [{pp.get('id')}]")
        if pp.get("painpoint_processed"):
            print(f"    -> {trunc(pp['painpoint_processed'], 200)}")
        if pp.get("feedback_id"):
            print(f"    feedback: {pp['feedback_id']}")


def cmd_signal_market(args) -> None:
    data = request(f"/v1/projects/{args.project}/signals/{args.id}/market-research")
    if args.json:
        return out_json(data)
    items = data.get("items", [])
    if not items:
        return print("no market research for this signal")
    print(f"{len(items)} market sources\n")
    for r in items:
        score = r.get("score")
        src = " / ".join(x for x in (r.get("company_name"), r.get("product_name")) if x)
        print(f"[{score:.3f}] {src or '(unknown source)'}" if score is not None
              else f"  {src or '(unknown source)'}")
        print(f"    {trunc(r.get('text'), 120)}")
        if r.get("business_description"):
            print(f"    about: {trunc(r['business_description'], 100)}")
        if r.get("released_at"):
            print(f"    released: {r['released_at']}")


# --------------------------------------------------------------------------- main
def build_parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(prog="ct.py", description="Continuum Tracker read-only API helper")
    sub = p.add_subparsers(dest="command", required=True)

    def add(name, fn, project=False):
        sp = sub.add_parser(name)
        sp.set_defaults(func=fn)
        sp.add_argument("--json", action="store_true", help="raw JSON output")
        if project:
            sp.add_argument("--project", required=True, help="project UUID")
        return sp

    add("me", cmd_me)
    add("projects", cmd_projects)
    add("project", cmd_project, project=True)   # single project, full mission/vision/north_star

    sp = add("signals", cmd_signals, project=True)
    sp.add_argument("--top", type=int, help="show only top N by feedback_count")
    sp.add_argument("--limit", type=int, help="page size (default 100)")
    sp.add_argument("--search")
    sp.add_argument("--category", action="append",
                    help="Recommendation|ManualPainPoint|ClusterPainPoint (repeatable)")
    sp.add_argument("--saved", action="store_true", help="only user-saved signals")
    sp.add_argument("--all", action="store_true", help="follow all pages")

    sp = add("feedbacks", cmd_feedbacks, project=True)
    sp.add_argument("--search")
    sp.add_argument("--limit", type=int)
    sp.add_argument("--all", action="store_true")

    sp = add("feedback", cmd_feedback, project=True)
    sp.add_argument("--id", required=True, help="feedback UUID")

    sp = add("signal", cmd_signal, project=True)
    sp.add_argument("--id", required=True, help="signal UUID")

    sp = add("signal-feedbacks", cmd_signal_feedbacks, project=True)
    sp.add_argument("--id", required=True, help="signal UUID")
    sp.add_argument("--limit", type=int)
    sp.add_argument("--all", action="store_true")

    sp = add("painpoints", cmd_painpoints, project=True)
    sp.add_argument("--feedback", help="feedback UUID")
    sp.add_argument("--signal", help="signal UUID")
    sp.add_argument("--limit", type=int)
    sp.add_argument("--all", action="store_true")

    sp = add("signal-market", cmd_signal_market, project=True)
    sp.add_argument("--id", required=True, help="signal UUID")

    return p


def main() -> None:
    args = build_parser().parse_args()
    global KEY
    KEY = load_key()
    args.func(args)


if __name__ == "__main__":
    main()
